### PR TITLE
implement LOG_EVERY_N

### PR DESCRIPTION
**Pull request by yidawang**

### CODE REVIEW

File: include/dmlc/logging.h [hunk 1]
```diff
@@ -187,8 +187,11 @@ DEFINE_CHECK_FUNC(_NE, !=)
 #define DLOG_IF(severity, condition) LOG_IF(severity, condition)
 #endif
 
-// Poor man version of LOG_EVERY_N
-#define LOG_EVERY_N(severity, n) LOG(severity)
+#define LOG_EVERY_N(severity, n) \
+  static int LOG_OCCURRENCES_MOD_N = 0; \
+    if (++LOG_OCCURRENCES_MOD_N > n) LOG_OCCURRENCES_MOD_N -= n; \
+    if (LOG_OCCURRENCES_MOD_N == 1) \
+      LOG_##severity.stream()
```
[the remark this answers is not in the document]
Comment (yidawang) [reply include/dmlc/logging.h]: Sure, we definitely can simplify it.  Thanks!

**Comment by yidawang** [comment]
The fundamental issue of this PR is that if this method will make the static variable defined multiple times when the header file is included by multiple cc files. I am wondering if `LOG_EVERY_N` is not directly used but only included in the header, will the `define` statement initialize the static variable?

**Comment by anirudh2290** [comment]
@yidawang The #define should not matter and the code will only be replaced in places where it is called. So if the header file is included from multiple cc files but not called anywhere it should be fine. If it is called in both cc files, the code that gets executed first (assuming both cc files are executed as part of same program lifetime) will initialize the static variable. 

One problem I see is for multithreaded scenario, where multiple threads may end up calling LOG_EVERY_N. I think thread_local should solve your problem http://en.cppreference.com/w/cpp/language/storage_duration but it is available only since C++11.

**Comment by tqchen** [comment]
The multiple definition place can be a problem, i suggest leave such macro for local use and not merge this in for now

**Comment by anirudh2290** [comment]
@tqchen can you explain how it can be a problem ?

**Comment by tqchen** [comment]
The major issue is when the static variable get defined in multiple cc files or even header files(if LOG_EVERY_N is used in another header). This will cause multiple def problem of static variable

**Comment by anirudh2290** [comment]
got it. thanks!

**Pull request closed without merging**